### PR TITLE
Markdownlint: `dashes-in-headings` custom rule

### LINKED ISSUE #28328: Markdownlint: Markdownlint and TOP web app generate different heading IDs/fragments
### Checks

- [X] This is not a duplicate of an existing issue (please have a look through our [open issues list](https://github.com/TheOdinProject/curriculum/issues) to make sure)
- [X] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
- [X] Would you like to work on this issue?

### Describe your suggestion

[Markdownlint's TOP051 rule](https://github.com/DavidAnson/markdownlint/blob/main/doc/md051.md) (valid link fragments) uses GitHub's heading algorithm for generating link fragments from headings. This preserves dashes, so a heading like `### Title - Subtitle` will generate `title---subtitle`.

TOP's web app's id/fragment generation removes these isolated dashes, so the above will become `foo-bar`. `#title---subtitle` will not scroll to the correct heading on our website but `#title-subtitle` will, except Markdownlint will flag `#title-subtitle` as an invalid link fragment as it expects `#title---subtitle` instead. This situation occurred recently in #28327.

A solution would be to replace dashes surrounded by spaces (so hyphenated words are preserved) with a colon and single space, since both serve to separate "parts" of a heading.

i.e. `### Title - subtitle` should be fixable to `### Title: Subtitle`.

## Acceptance criteria

- [ ] New markdownlint rule flags any level headings if they contain `SPACE DASH SPACE`, and suggest the colon equivalent in the error message
- [ ] Fix info should convert the heading to the colon equivalent
- [ ] Dashes used to hyphenate words must be left alone
- [ ] Test `.md` file should demonstrate all reasonable positive and negative cases
- [ ] The rule should have an appropriate doc file
- [ ] Rule added to `.markdownlint-cli2.jsonc` config file custom rules array

### Path

Other / NA

### Lesson Url

N/A

### (Optional) Discord Name

_No response_

### (Optional) Additional Comments

_No response_

### PR DESCRIPTION
**Pull request by MaoShizhong**
## Because

Markdownlint and VSC use [GitHub's heading algorithm](https://github.com/gjtorikian/html-pipeline/blob/f13a1534cb650ba17af400d1acd3a22c28004c09/lib/html/pipeline/toc_filter.rb) to generate fragments, which gives different results to [Ruby on Rails' `#parameterize` method](https://apidock.com/rails/String/parameterize) when dealing with isolated dashes (a dash group surrounded by spaces, used as a separator).

This leads to the TOP-valid fragment for `### Heading - Subheading` triggering the [MD051 rule](https://github.com/DavidAnson/markdownlint/blob/main/doc/md051.md), and the Markdownlint-valid fragment not matching the correct heading ID.

While we can use `### Heading - Subheading {#heading-subheading}` for Markdownlint and VSC to pick up, this looks messy and just gives us an extra point of failure if we change the heading but forget to change the custom ID.

The solution is to enforce colons as separators, leaving other dash uses (like hyphens or CLI flags) in headings alone. Markdownlint and our web app will both convert `### Heading: Subheading and hyphenated-word` to `heading-subheading-and-hyphenated-word`.

## This PR

- Adds TOP011 custom rule for dashes in headings
- Provides test file demonstrating what headings are flagged and not flagged by this custom rule
- Provides TOP011 doc file
- Adds TOP011 to the custom rules array in the markdownlint config file
- Amends TOP001-TOP010 doc files to use colon separators in headings for consistency

## Issue

Closes #28328

## Additional Information

This problem was encountered in [#28327](https://github.com/TheOdinProject/curriculum/pull/28327#discussion_r1661682283)

## Pull Request Requirements
-   [x] I have thoroughly read and understand [The Odin Project curriculum contributing guide](https://github.com/TheOdinProject/curriculum/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [x] The `Because` section summarizes the reason for this PR
-   [x] The `This PR` section has a bullet point list describing the changes in this PR
-   [x] If this PR addresses an open issue, it is linked in the `Issue` section
-   [ ] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [ ] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

### CODE REVIEW

File: markdownlint/docs/TOP011.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+```markdown
+### Headings - Hyphenated-words and --flags will be left alone
+
+<!-- Will get fixed to -->
+### Headings: Hyphenated-words and --flags will be left alone
+```
+
+## Rationale
+
+Markdownlint and Visual Studio Code uses the [GitHub heading algorithm](https://github.com/gjtorikian/html-pipeline/blob/f13a1534cb650ba17af400d1acd3a22c28004c09/lib/html/pipeline/toc_filter.rb) to generate the appropriate fragments from headings, which preserves *all* dashes in the heading. The TOP web app uses [Ruby on Rails' `#parameterize` method](https://apidock.com/rails/String/parameterize) to generate heading IDs and fragments, which removes separator dashes.
```
Comment (Asartea) [inline markdownlint/docs/TOP011.md:40]:
```suggestion
Markdownlint and Visual Studio Code use the [GitHub heading algorithm](https://github.com/gjtorikian/html-pipeline/blob/f13a1534cb650ba17af400d1acd3a22c28004c09/lib/html/pipeline/toc_filter.rb) to generate the appropriate fragments from headings, which preserves *all* dashes in the heading. The TOP web app uses the [Ruby on Rails `#parameterize` method](https://apidock.com/rails/String/parameterize) to generate heading IDs and fragments, which removes separator dashes.
```

Comment (MaoShizhong) [reply markdownlint/docs/TOP011.md:40]: Nice catch, and I'll take the fair nit

**Comment by MaoShizhong** [comment]
@KevinMulhern would appreciate your opinion here. Was just chatting to Eric about alternatives to implementing this rule, as I just realised this rule won't cover cases like `### This / that` being converted to `#this-that` by `#parameterize` in our app, and markdownlint using [Github's fragment conversion algo](https://github.com/gjtorikian/html-pipeline/blob/f13a1534cb650ba17af400d1acd3a22c28004c09/lib/html/pipeline/toc_filter.rb) which would convert it to `#thisthat`.

Was just wondering that as an alternative to this rule to cover only some of the mismatches in algos, whether we could nab the Github algo (adapted as necessary) and make a custom method in Rails to replace `#parameterize`. That'll sort out all discrepancies in link fragments between markdownlint and our website, without the need for a custom rule.

**Comment by MaoShizhong** [comment]
#28694 highlighted another issue with the discrepancy between MD051 using GitHub's section heading algo, and our website using `#parameterize`.

Closing since this PR will not be a sufficient fix - further discussion required.

**Pull request closed without merging**